lib/Runtime/Provisioner/Provisioner.cpp

```diff
@@ -143,82 +316,59 @@ Error Provisioner::provision(DAGListTy &networks, Module &module,
+            // reuse.
+            if (node->logicalDevices.size() > 1) {
+              functionMap.emplace(node->name, compiled.get());
+              duplicatedFunctions.emplace(node->name, std::move(compiled));
+              remainingDuplications[node->name] =
+                  node->logicalDevices.size() - 1;
+            } else {
+              compiledFunctions.emplace(node->name, std::move(compiled));
+            }
             break;
           }
         }
+        if (node->logicalDevices.size() == 1) {
+          functionMap.emplace(node->name, compiledFunctions[node->name].get());
+        }
       }
-      functionMap.emplace(node->name, compiledFunctions[node->name].get());
-      totalMemory += node->runtimeBundle->getConstantWeightSize();
     }
-    logicalDeviceSize.push_back(std::make_pair(device.first, totalMemory));
-    logicalDeviceBackendName[device.first] = nodeBackendName;
-    functionMaps.emplace(device.first, functionMap);
-  }
-  {
-    // Move compiled functions from compiledFunctions to functions_.
-    std::lock_guard<std::mutex> functionsLock(functionsLock_);
-    for (auto &func : compiledFunctions) {
-      functions_.emplace(func.first, std::move(func.second));
+    // Now that the functions are compiled add them to their assigned device
+    // then cleanup.
+    std::promise<void> addPromise;
+    auto ready = addPromise.get_future();
+    std::unique_ptr<Error> addErr;
+    devices_[physicalDevice]->addNetwork(
+        &module, functionMap,
+        [&addErr, &addPromise](const Module *, Error err) {
+          addErr = llvm::make_unique<Error>(std::move(err));
+          addPromise.set_value();
+        });
+    ready.wait();
+    DCHECK_NOTNULL(addErr.get());
+    if (*addErr.get()) {
+      cleanupProvision(localActiveNames);
+      return std::move(*addErr.get());
     }
-  }
-  // Sort by total size in descending order.
-  std::sort(logicalDeviceSize.begin(), logicalDeviceSize.end(), sortMostMemory);
-
-  // Get available memory for all devices.
-  std::vector<std::pair<DeviceIDTy, uint64_t>> deviceMemory;
-  for (unsigned i = 0; i < devices_.size(); i++) {
-    uint64_t availableMemory = devices_[i]->getAvailableMemory();
-    deviceMemory.push_back(std::make_pair(i, availableMemory));
-  }
-  // Sort by available memory in descending order.
-  std::sort(deviceMemory.begin(), deviceMemory.end(), sortMostMemory);
-
-  // Try to add functions to devices in order from largest to smallest.
-  std::map<std::string, size_t> startPos;
-  for (unsigned i = 0; i < logicalDeviceSize.size(); i++) {
-    std::string backendName =
-        logicalDeviceBackendName[logicalDeviceSize[i].first];
-    // Find the start point of each backendName device.
-    if (startPos.find(backendName) == startPos.end()) {
-      startPos[backendName] = 0;
+    // Free up memory no longer needed by the compiledFunction.
+    for (auto &func : functionMap) {
+      func.second->freeCompilationResources();
```

Comment:
Wouldn't this free up memory used by a duplicated function too (before we are done with it)?

Reply:
Good catch this was supposed to iterate over compiledFuncrtions.